File: src/main/java/ninekothecat/catconomy/commands/catconomycommand/CatEconomyCommandHandlerAutoCompleter.java
package ninekothecat.catconomy.commands.catconomycommand;

import ninekothecat.catconomy.Catconomy;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

public class CatEconomyCommandHandlerAutoCompleter implements TabCompleter {
    @Nullable
    @Override
    public List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String alias, @NotNull String[] args) {
        if (args.length == 1) {
            final Set<String> strings = new HashSet<>(Catconomy.catEconomyCommandHandler.getCommands().keySet());
            strings.removeIf(name -> !sender.hasPermission(Catconomy.catEconomyCommandHandler.getCommands().get(name).getPermission()));
            return new ArrayList<>(strings);
        } else if (args.length > 1 && Catconomy.catEconomyCommandHandler.getCommands().containsKey(args[0]) && Catconomy.catEconomyCommandHandler.get(args[0]).hasAutoComplete()) {
            List<String> arg2 = new ArrayList<>(Arrays.asList(args));
            arg2.remove(0);
            String[] args3 = arg2.toArray(new String[args.length - 1]);
            return Catconomy.catEconomyCommandHandler.get(args[0]).getAutoComplete().onTabComplete(sender, command, alias, args3);
        } else {
            return null;
        }

    }
}

File: src/main/kotlin/ninekothecat/catconomy/commands/catconomycommand/CatEconomyCommandHandlerAutoCompleter.kt
package ninekothecat.catconomy.commands.catconomycommand

import ninekothecat.catconomy.Catconomy
import org.bukkit.command.*
import java.util.*

class CatEconomyCommandHandlerAutoCompleter : TabCompleter {
    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        alias: String,
        args: Array<String>
    ): List<String>? {
        return if (args.size == 1) {
            val strings: MutableSet<String> =
                HashSet<String>(Catconomy.catEconomyCommandHandler!!.commands.keys)
            strings.removeIf { name: String ->
                !sender.hasPermission(
                    Catconomy.catEconomyCommandHandler!!.commands[name]!!.permission
                )
            }
            ArrayList(strings)
        } else if (args.size > 1 && Catconomy.catEconomyCommandHandler!!.commands
                .containsKey(args[0]) && Catconomy.catEconomyCommandHandler!![args[0]]!!
                .hasAutoComplete()
        ) {
            val arg2: MutableList<String> = ArrayList(listOf(*args))
            arg2.removeAt(0)
            val args3 = arg2.toTypedArray()
            Catconomy.catEconomyCommandHandler!![args[0]]!!.autoComplete!!.onTabComplete(sender, command, alias, args3)
        } else {
            null
        }
    }
}